jsonrpc/backend/tracer.go

```diff
@@ -44,33 +45,49 @@ func (b *JSONRPCBackend) TraceBlockByNumber(ethBlockNum rpc.BlockNumber, config
 		return nil, err
 	}
 
-	rpcTxs, err := b.getBlockTransactions(blockNumber)
+	bn := int64(blockNumber)
+	cosmosBlock, err := b.clientCtx.Client.Block(ctx, &bn)
 	if err != nil {
 		return nil, err
-	} else if len(rpcTxs) == 0 {
-		return nil, nil
 	}
 
 	var (
-		results = make([]*rpctypes.TxTraceResult, len(rpcTxs))
+		cosmosTxs = cosmosBlock.Block.Data.Txs
+		results   = make([]*rpctypes.TxTraceResult, 0, len(cosmosTxs))
 	)
 
+	idx := 0
 	sdkCtx := sdk.UnwrapSDKContext(ctx)
-	for i, rpcTx := range rpcTxs {
-		tx := rpcTx.ToTransaction()
+	for _, cosmosTxBytes := range cosmosTxs {
+		cosmosTxHash := cosmosTxBytes.Hash()
+		cosmosTx, err := b.app.TxDecode(cosmosTxBytes)
+		if err != nil {
+			return nil, err
+		}
+
+		// If the tx is not evm tx or cosmos tx which is not indexed, we need to run it without tracer
+		// and continue to the next tx
+		txHash, err := b.TxHashByCosmosTxHash(cosmosTxHash)
+		isCosmosTx := err == nil
```

Comment:
isn't it when txhash is nil?
```suggestion
		isCosmosTx := txHash == nil
```